File: cdk/NestedStacks/IpV6NetworkStack.cs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
using Amazon.CDK;
using Amazon.CDK.AWS.EC2;
using Amazon.AWSLabs.MultiAZWorkshop.Constructs;

namespace Amazon.AWSLabs.MultiAZWorkshop.NestedStacks
{
    public class IPV6NetworkStackProps : NestedStackProps, INestedStackProps
    {
        public string[] AvailabilityZoneNames {get; set; }
    }

    public class IpV6NetworkStack : NestedStackWithSource
    {
        public VpcIpV6 Vpc { get;}
        public string AvailabilityZoneNames {get;}

        public IpV6NetworkStack(Stack scope, string id, IPV6NetworkStackProps props) : base(scope, id, props)
        {
            this.Vpc = new VpcIpV6(this, "vpc", new VpcIpV6Props() {
                IpAddresses = IpAddresses.Cidr("192.168.0.0/16"),
                EnableDnsHostnames = true,
                EnableDnsSupport = true,
                CreateInternetGateway = false,
                AvailabilityZones = props.AvailabilityZoneNames,
                SubnetConfiguration = new ISubnetIpV6Configuration[] {
                    new SubnetIpV6Configuration() { CidrMask = 24, Name = "isolated-subnet", SubnetIpConfiguration = IPAddressType.IPv4, SubnetType = SubnetType.PRIVATE_ISOLATED},
                },
                RestrictDefaultSecurityGroup = false
            });

            this.Vpc.AddGatewayEndpoint("s3", new GatewayVpcEndpointOptions() { Service = GatewayVpcEndpointAwsService.S3 });
            this.Vpc.AddInterfaceEndpoint("vpcessm", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.SSM, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("vpcessmmessages", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.SSM_MESSAGES, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("kms", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.KMS, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("logs", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.CLOUDWATCH_LOGS, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("cloudwatch", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.CLOUDWATCH_MONITORING, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("ec2messages", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.EC2_MESSAGES, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("cfn", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.CLOUDFORMATION, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("xray", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.XRAY, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("codedeploy", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.CODEDEPLOY, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("codedeployagent", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.CODEDEPLOY_COMMANDS_SECURE, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("secretsmanager", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.SECRETS_MANAGER, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("elasticloadbalancing", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.ELASTIC_LOAD_BALANCING, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("sts", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.STS, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("ec2", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.EC2, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("ecrapi", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.ECR, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("ecrdkr", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.ECR_DOCKER, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("eks", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.EKS, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("eksauth", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.EKS_AUTH, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("autoscaling", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.AUTOSCALING, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("lambda", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.LAMBDA, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("stepfunctions", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.STEP_FUNCTIONS, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
            this.Vpc.AddInterfaceEndpoint("stepfunctionssync", new InterfaceVpcEndpointOptions() { Service = InterfaceVpcEndpointAwsService.STEP_FUNCTIONS_SYNC, Subnets = new SubnetSelection() { SubnetType = SubnetType.PRIVATE_ISOLATED}, PrivateDnsEnabled = true, Open = true } );
        }
    }
}

File: src/cdk/lib/nested-stacks/ipv6-network-stack.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { IPAddressType } from '../constructs/ip-address-type';
import { NestedStackWithSource } from '../constructs/nested-stack-with-source';
import { VpcIpV6, IVpcIpV6 } from '../constructs/vpc-ipv6-construct';

/**
 * Props for IPv6 Network Stack
 */
export interface IpV6NetworkStackProps extends cdk.NestedStackProps {
  /**
   * Availability zone names for the VPC
   */
  readonly availabilityZoneNames: string[];
}

/**
 * Nested stack that creates a VPC with IPv6 support and VPC endpoints
 */
export class IpV6NetworkStack extends NestedStackWithSource {
  /**
   * The VPC with IPv6 support
   */
  public readonly vpc: IVpcIpV6;

  /**
   * Availability zone names
   */
  public readonly availabilityZoneNames: string[];

  constructor(scope: cdk.Stack, id: string, props: IpV6NetworkStackProps) {
    super(scope, id, props);

    this.availabilityZoneNames = props.availabilityZoneNames;

    // Check if IPv6 is enabled via context
    const ipV6Enabled = scope.node.tryGetContext('ipV6Enabled') === true;

    // Determine IP address type based on context
    const subnetIpConfiguration = ipV6Enabled ? IPAddressType.DualStack : IPAddressType.IPv4;

    // Create VPC with IPv6 support
    this.vpc = new VpcIpV6(this, 'vpc', {
      ipAddresses: ec2.IpAddresses.cidr('192.168.0.0/16'),
      enableDnsHostnames: true,
      enableDnsSupport: true,
      createInternetGateway: false,
      availabilityZones: props.availabilityZoneNames,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'isolated-subnet',
          subnetIpConfiguration,
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
      ],
      restrictDefaultSecurityGroup: false,
    });

    // Add S3 Gateway Endpoint
    this.vpc.addGatewayEndpoint('s3', {
      service: ec2.GatewayVpcEndpointAwsService.S3,
    });

    // Add VPC Interface Endpoints
    this.vpc.addInterfaceEndpoint('vpcessm', {
      service: ec2.InterfaceVpcEndpointAwsService.SSM,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('vpcessmmessages', {
      service: ec2.InterfaceVpcEndpointAwsService.SSM_MESSAGES,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('kms', {
      service: ec2.InterfaceVpcEndpointAwsService.KMS,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('logs', {
      service: ec2.InterfaceVpcEndpointAwsService.CLOUDWATCH_LOGS,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('cloudwatch', {
      service: ec2.InterfaceVpcEndpointAwsService.CLOUDWATCH_MONITORING,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('ec2messages', {
      service: ec2.InterfaceVpcEndpointAwsService.EC2_MESSAGES,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('cfn', {
      service: ec2.InterfaceVpcEndpointAwsService.CLOUDFORMATION,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('xray', {
      service: ec2.InterfaceVpcEndpointAwsService.XRAY,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('codedeploy', {
      service: ec2.InterfaceVpcEndpointAwsService.CODEDEPLOY,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('codedeployagent', {
      service: ec2.InterfaceVpcEndpointAwsService.CODEDEPLOY_COMMANDS_SECURE,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('secretsmanager', {
      service: ec2.InterfaceVpcEndpointAwsService.SECRETS_MANAGER,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('elasticloadbalancing', {
      service: ec2.InterfaceVpcEndpointAwsService.ELASTIC_LOAD_BALANCING,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('sts', {
      service: ec2.InterfaceVpcEndpointAwsService.STS,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('ec2', {
      service: ec2.InterfaceVpcEndpointAwsService.EC2,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('ecrapi', {
      service: ec2.InterfaceVpcEndpointAwsService.ECR,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('ecrdkr', {
      service: ec2.InterfaceVpcEndpointAwsService.ECR_DOCKER,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('eks', {
      service: ec2.InterfaceVpcEndpointAwsService.EKS,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('eksauth', {
      service: ec2.InterfaceVpcEndpointAwsService.EKS_AUTH,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('autoscaling', {
      service: ec2.InterfaceVpcEndpointAwsService.AUTOSCALING,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('lambda', {
      service: ec2.InterfaceVpcEndpointAwsService.LAMBDA,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('stepfunctions', {
      service: ec2.InterfaceVpcEndpointAwsService.STEP_FUNCTIONS,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });

    this.vpc.addInterfaceEndpoint('stepfunctionssync', {
      service: ec2.InterfaceVpcEndpointAwsService.STEP_FUNCTIONS_SYNC,
      subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      privateDnsEnabled: true,
      open: true,
    });
  }
}
